pkg/operator/ceph/cluster/controller.go

```diff
@@ -242,15 +248,20 @@ func (c *ClusterController) onAdd(obj interface{}) {
 	}
 
 	cluster := newCluster(clusterObj, c.context, c.csiConfigMutex)
-	c.clusterMap[cluster.Namespace] = cluster
 
+	// Note that this lock is held through the callback process, as this creates CSI resources, but we must lock in
+	// this scope as the clusterMap is authoritative on cluster count and thus involved in the check for CSI resource
+	// deletion. If we ever add additional callback functions, we should tighten this lock.
+	c.csiConfigMutex.Lock()
```

Comment:
We may not actually need the lock. The CephCluster controller only processes a single CR event at a time. So if you add a new cluster CR, it won't be processed until the operator is done with other operations to add/update/delete a cluster CR. Or can you see another reason we would need the lock?

Reply:
I see. So you're saying that cluster.mon.mon.saveMonConfig may call SaveMonConfig in simultaneous goroutines, justifying the lock; but that all create, update, and delete actions on CephClusters are single-threaded.

I confess that my primary reason for these locks was against a faulty assumption that create and delete operations might be concurrent. If every operation fully completes before every other, I agree that this is largely unnecessary.

I guess I have a few questions as to whether it's still even vaguely worth it:

1) a) If CephCluster operations are single-threaded but mon.saveMonConfig is not, is there a possibility that mon.go could attempt to update the ClusterConfig mid-deletion (just as currently there is a possibility that it might update during another thread's update)?
1b) Since cluster_config.SaveClusterConfig is exported, is there a plausible risk that another caller of the function could fall into the same problem?
1c) Would we care in either of the above cases? (This is more a kube resource handling question than an us question, but still important.)

2) Are concurrent cluster CR operations a goal of the team in the short or mid term?

3) Against the possibility of any of the above, is there a reason *not* to lock here? Clearly these are extremely low-risk cases, but by the same token there's not a lot of reward in removing the lock.

All that being asked, though, my intention was to protect against concurrency issues between multiple creation and deletion operations. Since that's presently an impossibility and premature code is often the enemy, I'm very happy to remove the locks. Let me know when you get a chance. Thanks!

Reply:
Good questions. While concurrent actions with the CRs are not possible (and I expect it to stay this way), you raise a good point about the configmap being updated from another goroutine. The other goroutine is ensuring the mons are healthy. If a mon fails, the mon will be failed over and the configmap will be updated with the new mon endpoint. There is a lock so that this goroutine and the CR controller won't attempt to orchestrate the mons at the same time. It does seem like we should use this same lock as well for deleting or accessing the csi configmap. Instead of creating a new lock, can you see if we can use the same lock that already exists for the mons?  Or is the lock you created protecting the `SaveClusterConfig()` method from this configmap deletion already?

Reply:
It was certainly my intention to use the same mutex, and I believe it is in this patch. The call sequence to verify this is:

- In ClusterController.onAdd, just before we lock to create the config, we create the new cluster: controller.go#250:
`cluster := newCluster(clusterObj, c.context, c.csiConfigMutex)`
- newCluster passes this mutex on to mon.New: cluster.go#88: 
`mons: mon.New(context, c.Namespace, c.Spec.DataDirHostPath, c.Spec.Network, ownerRef, csiMutex, false)`
- This is stored in the new Cluster struct as csiConfigMutex. Cluster.saveMonConfig then passes this to csi.SaveClusterConfig: mon.go#768: 
`if err := csi.SaveClusterConfig(c.context.Clientset, c.Namespace, c.ClusterInfo, c.csiConfigMutex); err != nil {`

We use this same mutex to lock the ConfigMap creation directly after creating the new cluster at controller.go#255. It's a little convoluted, but it's the same lock. (Or at very least, if it's not then the mons are not using the same lock as one another and we need to rethink the entire synchronization mechanism.)

Reply:
Ok, I see now that `csiConfigMutex` already exists. For some reason I was thinking you just created it. I was thinking of another mutex `orchestrationMutex` and wondering if we could consolidate to use just that one instead of csiConfigMutex. But we can save that question for another day. For now, sounds good to use this mutex as you have it here.

Reply:
ochestrationMutext seems like a different beast. To my inspection it seems that csiConfigMutext is intended to be both a bit more surgical than orchestrationMutex (as it locks only csiConfig changes) and broader in scope (as it is local to a cluster, while csiConfigMutex locks across all clusters). It might be best to leave them as-is.

Reply:
Sounds good to leave them separate. `orchestrationMutex` has a much bigger scope.